mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp

```diff
@@ -444,7 +444,15 @@ scf::ForOp LoopPipelinerInternal::createKernelLoop(
         loc, rewriter.getIntegerAttr(t, maxStage));
     Value maxStageByStep =
         rewriter.create<arith::MulIOp>(loc, step, maxStageValue);
-    newUb = rewriter.create<arith::SubIOp>(loc, ub, maxStageByStep);
+    Value hasAtLeastOneIteration = rewriter.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::slt, maxStageByStep, ub);
```

Comment:
Actualy what we want to make sure is that `maxStageByStep < ub`, because if those value are from `unsigned` or `index` may lead to wrong answer. Please see full answer in the second thread.